### PR TITLE
Testing section content

### LINKED ISSUE #38: Content Contribution: Testing page
## Description

Add content to the Development/Testing section of the guide.

**Filename:** `/guide/source/dev/testing.md`

**Page/Section:** Development/Testing

**Contributing Author/Company**: Ben Kehoe (iRobot)

**Details:** 

Some suggested ideas and guidance for content in this section:

#### Introduction

[Brief intro into unit testing and integration testing]

#### Unit Testing

Adapt from  [https://serverless.zone/unit-and-integration-testing-for-lambda-fc9510963003]

- How unit testing in serverless needs a new perspective?
- Why abstracting out service invocations is not optimal?
- How can mocking / stubbing calls help?
- Any challenges?

#### Integration Testing

Adapt from [https://serverless.zone/monkeyless-chaos-integration-testing-for-serverless-architectures-9d1a4208c3ca]

- Challenges in writing integration tests?
- Integration testing: Local vs Live
- Testing that all functions for a service work in cohesion as part of a larger workflow
- Testing that the system is resilient to degradation or stability of dependent services
- Monkeyless Chaos
  - Describe the pattern, its pieces, maybe a diagram
  - Requirements
  - Suggestions for building such a system

### PR DESCRIPTION
**Pull request by benkehoe**
## What did you implement:
imported content for Testing section from Medium articles

Closes #38

## Todos:

- [x] Have you checked for typos and grammatical errors?
- [x] If applicable, have you updated the [TOC](/guide/source/README.md)?

**Is this ready for review?:** [Yes/No]

### CODE REVIEW

File: source/dev/testing.md [hunk 1]
```diff
@@ -9,4 +9,114 @@ publish: false
...
+                'placebo')
+            pill = placebo.attach(session, data_path=path)
+            return session
+        Boto3Wrapper.SESSION_CREATE_HOOK = attach_placebo
+    def test_function_requirement_1(self):
+        # perform test, Lambda function will automatically get
+        # placebo injected on its sessions
+```
+
+This approach for functions to be written as concisely as possible, focusing on business logic, and letting abstraction take place at the architecture level, in the separation of code and APIs between functions.
```
Comment (rupakg) [inline source/dev/testing.md]: Minor typo: This approach **allows** for functions

File: source/dev/testing.md [hunk 2]
```diff
@@ -9,4 +9,114 @@ publish: false
...
+
+There’s one caveat here: the way boto3 works, to set up the intercept, you have to call a method on the session that is being used by a given client or resource, and there is no hook provided at the package level to inject this into the Session constructor. An unrelated detail: boto3 sessions, clients, and resources are relatively expensive to create. For these two reasons, you can use one minor abstraction across your functions: create a class that provides factory methods for sessions, clients, and resources. This class should provide a hook for injecting placebo into the session, and also cache the sessions, clients, and resources, which will then reduce the overhead across successive function invocations. It could look more or less like this:
+
```
Comment (rupakg) [inline source/dev/testing.md]: Simplifying a bit:

Creating an abstraction (`Boto3Wrapper` class) that provides factory methods for sessions, clients, and resources will enable caching, which will then reduce the overhead across successive function invocations. For example:

File: source/dev/testing.md [hunk 3]
```diff
@@ -9,4 +9,114 @@ publish: false
...
...system, abstraction isn’t worth it for getting around the level of vendor lock-in that comes with these designs; a lowest-common denominator interface that can talk to both, say, AWS DynamoDB and Google Cloud Bigtable is going to have limited functionality with basically no opportunities to take advantage of either service’s optimization techniques. Two, abstracting the service invocations for the purposes of testing is unnecessary, and, in fact, creates extra work! The AWS SDK provides mechanisms for stubbing out SDK calls. There are options for the JavaScript SDK in [aws-sdk-mock](https://www.npmjs.com/package/aws-sdk-mock) and [mock-aws](https://www.npmjs.com/package/mock-aws), but if you code in Python, you can use [placebo](https://github.com/garnaat/placebo) (though there is also [moto](https://github.com/spulec/moto)). With placebo, you passively record SDK calls on a real session, and then for testing you can instruct boto3 to use the recorded response instead of actually making the call. So your function can happily proceed to directly use the SDK (without abstractions), and it’ll never know that it’s inside a testing environment, not actually talking to the outside world.
```
Comment (rupakg) [inline source/dev/testing.md]: Just breaking up the big paragraph for readability:

A serverless function, by definition, can only have side effects by using other services. Unlike in a traditional, non-serverless system, it’s less necessary to abstract out the service invocations, for two reasons:

- One, in a serverless/SaaS-based system, abstraction isn’t worth it for getting around the level of vendor lock-in that comes with these designs; a lowest-common denominator interface that can talk to both, say, AWS DynamoDB and Google Cloud Bigtable is going to have limited functionality with basically no opportunities to take advantage of either service’s optimization techniques.
- Two, abstracting the service invocations for the purposes of testing is unnecessary, and, in fact, creates extra work! The AWS SDK provides mechanisms for stubbing out SDK calls. Using recorded responses as a mocking technique, the function can be tricked to believe it is making the live call. No abstractions are required. 
> There are several mocking libraries - [aws-sdk-mock](https://www.npmjs.com/package/aws-sdk-mock) & [mock-aws](https://www.npmjs.com/package/mock-aws) for JS and [placebo](https://github.com/garnaat/placebo) & [moto](https://github.com/spulec/moto) for Python.

File: source/dev/testing.md [hunk 4]
```diff
@@ -9,4 +9,114 @@ publish: false
 
 ### Testing
 
-To contribute to this section of the guide, please see suggested content breakup in [Issue #38](https://github.com/serverless/guide/issues/38)
\ No newline at end of file
+#### Introduction
+
+[Brief intro into unit testing and integration testing]
+
```
Comment (rupakg) [inline source/dev/testing.md]: Maybe these two lines from the para below can be part of the introduction:

When thinking about testing, it's useful with serverless functions to think in terms of unit tests that are performed against a function in isolation, and integration tests that test the system as a whole.

File: source/dev/testing.md [hunk 5]
```diff
@@ -9,4 +9,114 @@ publish: false
...
+#### Unit Testing
+
+Lambda functions are ideally small — a few hundred lines of code at the most — taken up mostly by error handling; the happy path should be very short, or at least relatively straightforward. Thus, introducing abstractions can create a lot of code bloat. 
+
+When thinking about testing, it's useful with serverless functions to think in terms of unit tests that are performed against a function in isolation, and integration tests that test the system as a whole. Unit tests can be performed locally, because they just require the code for the function, but integration tests involving SaaS can really only be performed on the deployed system (some platforms may offer local executions as part of a deployed system, which mitigates this somewhat). So what should serverless function unit tests look like? For an example, we'll use the AWS Python SDK, [boto3](https://boto3.readthedocs.io/en/latest/).
+
```
Comment (rupakg) [inline source/dev/testing.md]: Lambda functions should ideally be small — a few hundred lines of code at the most — taken up mostly by error handling. The happy path should be very short, or at least relatively straightforward. Thus, introducing abstractions can create a lot of code bloat. 

So what should unit tests for serverless functions look like? 

> We'll use the AWS Python SDK, [boto3](https://boto3.readthedocs.io/en/latest/) in our examples.

File: source/dev/testing.md (hunk 3 above)
Comment (andrea-p) [inline source/dev/testing.md]: made some edits for concision/voice

File: source/dev/testing.md (hunk 5 above)
Comment (andrea-p) [inline source/dev/testing.md]: cut out just a few words!

**Pull request merged**